### PR TITLE
Adding segmentation to DFM

### PR DESCRIPTION
**Pull request by nahuja-intel**
# Description

- Provide a summary of the modification as well as the issue that has been resolved. List any dependencies that this modification necessitates.

- Fixes # (issue)

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]
Minor comments but rest looks good. Thanks for the efforts!

File: anomalib/models/dfm/torch_model.py [hunk 1]
```diff
@@ -109,10 +113,11 @@ def fit(self, dataset: Tensor) -> None:
...
 
         self.pca_model.fit(dataset)
-        features_reduced = self.pca_model.transform(dataset)
-        self.gaussian_model.fit(features_reduced.T)
+        if self.score_type == 'nll':
+            features_reduced = self.pca_model.transform(dataset)
+            self.gaussian_model.fit(features_reduced.T)
 
-    def score(self, features: Tensor) -> Tensor:
+    def score(self, features: Tensor, feature_shape) -> Tensor:
```
[the remark this answers is not in the document]
Comment (djdameln) [reply anomalib/models/dfm/torch_model.py:120]: And maybe rename it to `feature_shapes` to be consistent with the naming elsewhere in the class.

File: anomalib/models/dfm/torch_model.py [hunk 2]
```diff
@@ -129,11 +134,19 @@ def score(self, features: Tensor) -> Tensor:
             score = self.gaussian_model.score_samples(feats_projected)
         elif self.score_type == "fre":
```
Comment (djdameln) [inline anomalib/models/dfm/torch_model.py:135]: If I undertand correctly, segmentation is only supported when score_type is "fre"? Maybe you could mention this in the docstring of the method, and also in `anomalib/models/dfm/README.md`

**Review by djdameln — CHANGES_REQUESTED** [review]
Thanks, very nice addition! I just have some minor comments

**Comment by nahuja-intel** [comment]
So, all the pre-commit checks, including isort, are passing when I commit locally. However isort is failing while requesting a PR. Any suggestions?

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks. One comment

File: anomalib/models/dfm/README.md [hunk 3]
```diff
@@ -14,7 +14,7 @@ Features are extracted by feeding the images through a ResNet18 backbone, which
 
 ### Anomaly Detection
 
-In the anomaly classification stage, class-conditional PCA transformations and Gaussian Density models are learned. Two types of scores are calculated (i) Feature-reconstruction scores (norm of the difference between the high-dimensional pre-image of a reduced dimension feature and the original high-dimensional feature), and (ii) Negative log-likelihood under the learnt density models. Either of these scores can be used for anomaly detection.
+In the anomaly classification stage, class-conditional PCA transformations and Gaussian Density models are learned. Two types of scores are calculated (i) Feature-reconstruction scores (norm of the difference between the high-dimensional pre-image of a reduced dimension feature and the original high-dimensional feature), and (ii) Negative log-likelihood under the learnt density models. Anomaly map generation is supported only with the feature-reconstruction based scores. Image level anomaly detection is supported by score types.
```
[the remark this answers is not in the document]
Comment (nahuja-intel) [reply anomalib/models/dfm/README.md:17]: Thanks for catching that. Yes, it should be "...supported by both score types". Any feedback on why the tox run failed? Anything I need to change in the code?

File: anomalib/models/dfm/lightning_model.py [hunk 4]
```diff
@@ -100,9 +104,9 @@ def validation_step(self, batch, _):  # pylint: disable=arguments-differ
           batch (List[Dict[str, Any]]): Input batch
 
         Returns:
-          Dictionary containing FRE anomaly scores and ground-truth.
+          Dictionary containing FRE anomaly scores and anomaly maps.
         """
-        batch["pred_scores"] = self.model(batch["image"])
+        batch["pred_scores"], batch["anomaly_maps"] = self.model(batch["image"])
```
Comment (ashwinvaidya17) [inline anomalib/models/dfm/lightning_model.py:109]: This needs to be assigned based on the score type. If I select `nll` score type then training fails.

**Review by ashwinvaidya17 — CHANGES_REQUESTED** [review]

File: anomalib/models/dfm/torch_model.py [hunk 5]
```diff
@@ -129,11 +137,19 @@ def score(self, features: Tensor) -> Tensor:
...
+            score_map = F.interpolate(fre_map, size=self.input_size, mode="bilinear", align_corners=False)
             score = torch.sum(torch.square(features - feats_reconstructed), dim=1)
         else:
             raise ValueError(f"unsupported score type: {self.score_type}")
 
-        return score
+        if self.score_type == "nll":
+            output = score
+        else:
+            output = score, score_map
```
Comment (djdameln) [inline anomalib/models/dfm/torch_model.py:150]: The torch inferencer expects the predictions of the torch model for the segmentation task type to be in the form (anomaly_map, anomaly_score), so these two should be reversed here (of course, the lightning model should then also be updated accordingly. 

This is what's causing the failing test case in the CI.

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks for the efforts

**Review by samet-akcay — APPROVED** [review]
lgtm, thanks!

**Review by djdameln — APPROVED** [review]

**Pull request merged**